src/headwordsmodel.cc

```diff
@@ -80,21 +80,19 @@ void HeadwordListModel::appendWord( const QString & word )
 void HeadwordListModel::requestFinished()
 {
   // See how many new requests have finished, and if we have any new results
-  for ( auto i = queuedRequests.begin(); i != queuedRequests.end(); ) {
-    if ( ( *i )->isFinished() ) {
-      if ( !( *i )->getErrorString().isEmpty() ) {
-        qDebug() << "error:" << ( *i )->getErrorString();
+  auto snapshot = queuedRequests.snapshot();
```

Comment:
The type of `queuedRequest` is `list< sptr< request > >`

The `snapshot` only copies the `sptr` and not the object that `sptr` is pointing to.

If `request` changes during the process, would it be crash?

(BTW, sometimes, it seems not using `auto` is better.)

Reply:
The constructor of `request` contains a `QFuture f = QConcurrent::run([](){})` that starts its own thread.

`request` doesn't have a copy constructor/assigment.

If we copy `request`, will C++ uses the user decleared constructor? Does this mean `request` cannot be reasonably copied? 😅 I have no idea.

There is a cursed diagram from https://howardhinnant.github.io/classdecl.html

<img width="400px" src="https://github.com/user-attachments/assets/71066f92-701b-4b0e-98ce-9076c4506ee4">

Reply:
To protect both the pointer and the object, Rust has an obvious solution `Arc<Mutex<T>>`. 

Arc is almost std::shared_ptr.

There is no obvious equivalent of `Mutex<T>`

By some googling, there appears to be some funny way to simulate the  technique.

https://stackoverflow.com/questions/67993098/how-to-simulate-rusts-mutexobject-in-c


Do Qt has something that reinvents the same pattern?

Reply:
the `request.use_count` will be increased.   the underline object is the same.

Reply:
headwordmodel seems do not need queuedRequests, I just remove it.